varname/ignore.py

```diff
@@ -176,42 +172,56 @@ def __init__(self, ignore: Tuple[ModuleType, str]) -> None:
         # check uniqueness of qualname
         self.qname_checked = False
         modfile = getattr(self.ignore[0], '__file__', None)
-        if modfile and path.isfile(modfile):
+        if modfile:
             source = Source.for_filename(modfile, self.ignore[0].__dict__)
-            nobj = list(source._qualnames.values()).count(self.ignore[1])
-            # don't check for non-existence, as it could contain wildcards
-            assert nobj <= 1, (
+            self._check_qualname_by_source(source)
+            self.qname_checked = True
+
+    @staticmethod
+    @lru_cache()
+    def _source_from_frame(frame: FrameType) -> Source:
+        """Get source object from the given frame"""
+        return Source.for_frame(frame)
+
+    def _check_qualname_by_source(self, source: Source) -> None:
+        """Check if qualname references to multiple objects"""
+        if not source.tree:
+            # there is no way to check
+            return
+        nobj = list(source._qualnames.values()).count(self.ignore[1])
+        # don't check for non-existence, as it could contain wildcards
+        if nobj > 1:
+            raise QualnameNonUniqueError(
                 f"Qualname {self.ignore[1]!r} in "
                 f"{self.ignore[0].__name__!r} refers to multiple objects."
             )
-            self.qname_checked = True
+
+    @lru_cache()
+    def _check_qualname_by_frame(self, frame: FrameType) -> None:
+        """Check if qualname references to multiple objects by given frame"""
+        source = IgnoreModuleQualname._source_from_frame(frame)
+        self._check_qualname_by_source(source)
 
     def match(self, frame_no: int, frameinfos: List[inspect.FrameInfo]) -> bool:
         frame = frameinfos[frame_no].frame
-
-        module = inspect.getmodule(frame)
+        module = cached_getmodule(frame.f_code)
         # Return earlier to avoid qualname uniqueness check
         if module and module != self.ignore[0]:
             return False
 
         if not module and (
-                getattr(self.ignore, MODULE_IGNORE_ID_NAME, None) !=
-                frame.f_globals.get(MODULE_IGNORE_ID_NAME, '')
+                getattr(self.ignore[0], MODULE_IGNORE_ID_NAME, object()) !=
```

Comment:
Well that was almost a disaster! I think to be safe and readable you should replace `self.ignore[0/1]` with actual named attributes, maybe even `del self.ignore`.